### PR TITLE
GDT: Retry uploading batches failed to upload before the app termination

### PR DESCRIPTION
**Pull request by maksymmalyhin**
Fixes b/159990839.

This PR fixes the specified issues as well as a couple other cases related to uploading discovered during testing. The following changes where applied:
- `GDTCCTUploader`:
  - **remove any batches left from previous app launches moving the events from the batch back to the main storage, so the events can be uploaded again on the next attempt** (the core fix for b/159990839)
  - fix race condition with tracking upload state (solved by maintaining `isCurrentlyUploading` atomic property)
  - refactor `dispatch_semaphore_t` queue locking to async methods to reduce CPU usage and risks of a deadlock
  - `uploadTarget:withConditions:` refactored to a number of smaller methods in an attempt to improve readability (still not ideal, b/160018519 created with ideas for further improvements)
  - test coverage increased
- `GDTCORFlatFileStorage`:
  - `removeBatchWithID:deleteEvents:onComplete` method fixed
  - tests
- various test utils changes to support required tests

### CODE REVIEW

**Comment by maksymmalyhin** [comment]
> Nothing obviously wrong to me. However, overall it's too much for me to follow. Perhaps this is worth a GVC to talk through?

Thank you for the review! Sure, GVC SG. Also, I tried to give meaningful commit names, so per-commit review may provide a bit better context.

File: GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m [hunk 1]
```diff
@@ -209,121 +225,205 @@ - (void)uploadTarget:(GDTCORTarget)target withConditions:(GDTCORUploadConditions
...
+    case kGDTCORTargetFLL:
+      // Falls through.
+    case kGDTCORTargetCSH:
+      self->_FLLNextUploadTime = futureUploadTime;
+    default:
+      break;
+  }
+
+  // Only retry if one of these codes is returned, or there was an error.
+  if (error || ((NSHTTPURLResponse *)response).statusCode == 429 ||
```
Comment (maksymmalyhin) [inline GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m:358]: TODO: Validate why those specific codes. + comments.

File: GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m [hunk 2]
```diff
@@ -439,6 +471,66 @@ - (void)storageSizeWithCallback:(void (^)(uint64_t storageSize))onComplete {
   });
 }
 
+#pragma mark - Private not thread safe methods
+
+- (nullable NSArray<NSString *> *)batchDirPatchPathsForBatchID:(NSNumber *)batchID
```
Comment (mikehaney24) [inline GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m:476]: I couldn't find docstring for this method.

File: GoogleDataTransport/GDTCORLibrary/Public/GDTCORStorageProtocol.h [hunk 3]
```diff
@@ -76,6 +76,11 @@ NS_ASSUME_NONNULL_BEGIN
 - (void)batchIDsForTarget:(GDTCORTarget)target
                onComplete:(void (^)(NSSet<NSNumber *> *_Nullable batchIDs))onComplete;
 
+/** Retrieves events for a batch with the specified ID. If a batch with ID is not found,  passes
```
Comment (mikehaney24) [inline GoogleDataTransport/GDTCORLibrary/Public/GDTCORStorageProtocol.h:79]: there are two spaces before 'passes'

**Comment by maksymmalyhin** [comment]
@mikehaney24 Thank you for the review! I'll address the comments.

**Comment by maksymmalyhin** [comment]
@paulb777 @samedson I updated the PR with a slightly different approach to retry. Instead of uploading the failed batch directly the batched events are moved back to the queue and then are batched according to the priorities. It allows to make the code a bit simpler.

The tests updates and cleanup is still TBD. Could you please do another review round.

File: GoogleDataTransport/GDTCCTTests/Unit/GDTCCTUploaderTest.m [hunk 4]
```diff
@@ -1,88 +1,482 @@
...
-    XCTAssertNotNil(uploader.currentTask);
-  });
-  [self waitForExpectations:@[ responseSentExpectation ] timeout:30.0];
-  dispatch_sync(uploader.uploaderQueue, ^{
-    XCTAssertNil(uploader.currentTask);
-  });
-}
-
-@end
+///*
```
Comment (samedson) [inline GoogleDataTransport/GDTCCTTests/Unit/GDTCCTUploaderTest.m:1]: Does this need to be uncommented?

Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCCTTests/Unit/GDTCCTUploaderTest.m:1]: Yes, these tests are to be uncommented and updated.

File: GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m [hunk 5]
```diff
@@ -539,6 +539,40 @@ - (void)storageSizeWithCallback:(void (^)(uint64_t storageSize))onComplete {
   return [eventPathsForBatchID copy];
 }
 
+/** Makes a copy of the contents of a directory to a directory at the specified path.*/
+- (BOOL)moveContentsOfDirectoryAtPath:(NSString *)sourcePath
```
Comment (paulb777) [inline GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m:547]: Why doesn't `moveItem` meet the requirements - https://developer.apple.com/documentation/foundation/filemanager/1413529-moveitem

Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m:561]: Added clarification comments to the code. The following `moveItem` behaviour makes it failing for the case:
```
...If an item with the same name already exists at dstPath, this method stops the move attempt and returns an appropriate error...
```

**Comment by maksymmalyhin** [comment]
Based on my testing the current version works correctly. 

@samedson @visumickey Taking into account the complexity of the logic, I would like to ask you to help with testing by validating Crashlytics and Performance SDKs with this GDT version. Thanks.

File: GoogleDataTransport/GDTCORLibrary/Public/GDTCORStorageProtocol.h [hunk 6]
```diff
@@ -25,6 +25,9 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
+typedef void (^GDTCORStorageBatchBlock)(NSNumber *_Nullable newBatchID,
+                                        NSSet<GDTCOREvent *> *_Nullable batchEvents);
+
```
[the remark this answers is not in the document]
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCORLibrary/Public/GDTCORStorageProtocol.h:30]: Ah, I din't know this, thank you for the heads up!

File: GoogleDataTransport.podspec [hunk 7]
```diff
@@ -125,13 +125,13 @@ Shared library for iOS SDK data transport needs.
...
 
-  common_test_sources = ['GoogleDataTransport/GDTCCTTests/Common/**/*.{h,m}']
+  common_cct_test_sources = ['GoogleDataTransport/GDTCCTTests/Common/**/*.{h,m}']
 
   # Test specs
   s.test_spec 'CCT-Tests-Unit' do |test_spec|
     test_spec.platforms = {:ios => '8.0', :osx => '10.11', :tvos => '10.0'}
     test_spec.requires_app_host = false
-    test_spec.source_files = ['GoogleDataTransport/GDTCCTTests/Unit/**/*.{h,m}'] + common_test_sources
+    test_spec.source_files = ['GoogleDataTransport/GDTCCTTests/Unit/**/*.{h,m}'] + common_cct_test_sources + common_test_sources
```
[the remark this answers is not in the document]
Comment (maksymmalyhin) [reply GoogleDataTransport.podspec:134]: Yes, it is defined [here](https://github.com/firebase/firebase-ios-sdk/blob/928534437b4bec81d1c0f22741fcd2fdd3359221/GoogleDataTransport.podspec#L55)

File: GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m [hunk 8]
```diff
@@ -190,210 +206,341 @@ - (NSString *)FLLAndCSHAPIKey {
...
   } else {
     GDTCORLogDebug(@"CCT: can't upload to target %ld because the backend asked to wait",
                    (long)target);
   }
-  result = YES;
+
+  isAfterNextUploadTime = YES;
   GDTCORLogDebug(@"CCT: can upload to target %ld because nothing is preventing it", (long)target);
-  return result;
+  return isAfterNextUploadTime;
```
[the remark this answers is not in the document]
Comment (paulb777) [reply GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m:543]: But what about the case where isAfterNextUploadTime was NO in the last test?

File: GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m [hunk 9]
```diff
@@ -190,210 +206,341 @@ - (NSString *)FLLAndCSHAPIKey {
...
+
+                          return;
+                        }
+                        // 3. Upload events.
+                        [self uploadBatchWithID:batchID
+                                         events:events
+                                         target:target
+                                        storage:storage
+                                     completion:^{
+                                       backgroundTaskCompletion();
```
[the remark this answers is not in the document]
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m:258]: Yeah, it's handled in the method itself. Agree, the readability and using `isCurrentlyUploading` approach in  can be improved in general, but it will require bigger changes which I would prefer to implement in followup PRs.

File: GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m (hunk 8 above)
[the remark this answers is not in the document]
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m:543]: Good catch, this past won't work as expected. Let me fix it.

File: GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m [hunk 10]
```diff
@@ -190,210 +206,341 @@ - (NSString *)FLLAndCSHAPIKey {
...
-#if !NDEBUG
-      // Post a notification when in DEBUG mode to state how many packages were uploaded. Useful
-      // for validation during tests.
-      [[NSNotificationCenter defaultCenter] postNotificationName:GDTCCTUploadCompleteNotification
-                                                          object:@(events.count)];
-#endif  // #if !NDEBUG
+    case kGDTCORTargetFLL:
+      // Falls through.
+    case kGDTCORTargetCSH:
+      self->_FLLNextUploadTime = futureUploadTime;
```
[the remark this answers is not in the document]
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m:368]: Good catch!

File: GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m (hunk 9 above)
[the remark this answers is not in the document]
Comment (visumickey) [reply GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m:258]: I agree. It took sometime to understand how this state property is managed.

Not sure if this could be a possible step forward here: can we have `isCurrentlyUploading = YES` in the step 3 above. This would atleast reduce to confusion of hopping different areas of the code to see how this state is managed. Just a nit!

**Comment by visumickey** [comment]
Thanks for making the fix. This definitely seems more involved and so, great work.

Since the change by itself is quite big, I have a recommendation for future changes of this size. May be creating a separate master branch for "gdtRetryFixes" and make separate PRs for every change separately such "gdtRetryFixesAddUploadingState" and have them merged with "gdtRetryFixes". Once finally all the changes are done, create another PR to merge with master?

This change on its entirety is huge and sometimes hard to follow. Might be helpful to break it down (Not now, in future for other changes).

**Comment by maksymmalyhin** [comment]
@visumickey Thank you for taking a look and the suggestions! Sorry for the big PR. I completely agree smaller PRs would be nicer to review. The nature of the issue and fixes made following this approach productively in timely manner pretty difficult because the code was not really working until most of the issue in the PR were fixed. 

I tried to give meaningful scopes and name for commits, so you may consider most of them like a separate small PR to be reviewed one by one.

File: GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m (hunk 9 above)
[the remark this answers is not in the document]
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m:258]: Unfortunately setting `isCurrentlyUploading = YES` in the step 3 is already a bit late and may lead to race condition. 

In the future I would like to refactor the class to use Promises that will allow to express the sequence more elaborative and will provide a way to manage the state i a single place. It is expected to be updated in the scope of b/160018519.

File: GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m [hunk 11]
```diff
@@ -190,210 +206,341 @@ - (NSString *)FLLAndCSHAPIKey {
...
-      if (bgID != GDTCORBackgroundIdentifierInvalid) {
-        [[GDTCORApplication sharedApplication] endBackgroundTask:bgID];
-        bgID = GDTCORBackgroundIdentifierInvalid;
-      }
-      self.currentTask = nil;
-    };
+  // Only retry if one of these codes is returned, or there was an error.
+  if (error || ((NSHTTPURLResponse *)response).statusCode == 429 ||
+      ((NSHTTPURLResponse *)response).statusCode == 503) {
+    [storage removeBatchWithID:batchID deleteEvents:NO onComplete:nil];
```
[the remark this answers is not in the document]
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m:378]: Comment added.

File: GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m [hunk 12]
```diff
@@ -190,210 +206,341 @@ - (NSString *)FLLAndCSHAPIKey {
...
-                     @"An upload shouldn't be initiated with another in progress.");
+  if (self.isCurrentlyUploading) {
+    GDTCORLogDebug(@"%@", @"CCT: Wait until previous upload finishes. The current version supports "
+                          @"only a single batch uploading at the time.");
     return NO;
   }
 
-  if (batchIDs.count > 0) {
-    GDTCORLogDebug(@"%@", @"CCT: can't upload because a package is in flight");
+  // Not ready to upload with not network connection.
```
[the remark this answers is not in the document]
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m]: fixed

File: GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m (hunk 10 above)
[the remark this answers is not in the document]
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m:368]: fixed

File: GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m (hunk 8 above)
[the remark this answers is not in the document]
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m]: Added tests for the logic implemented in the method and fixed the implementation itself (see the [commits](https://github.com/firebase/firebase-ios-sdk/pull/5942/files/0c33e27c6bc31128724bc4b845bb9fed738091c8..d94fc9b8ce44e1e8e8fba7cfb92c61d543bc4f27)).

File: GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m (hunk 1 above)
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m:375]: This will be addressed in a separate PR

**Comment by maksymmalyhin** [comment]
I'm merging the PR. Feel free to add comments if needed. They will be addressed in the next PRs.
